pyquil/api/_quantum_computer.py

```diff
@@ -671,3 +695,340 @@ def local_qvm() -> Iterator[Tuple[subprocess.Popen, subprocess.Popen]]:
+        prog_symm = _flip_array_to_prog(flip_array, meas_qubits)
+        total_prog_symm += prog_symm
+        symm_programs.append(total_prog_symm)
+        flip_arrays.append(flip_array)
+
+    # ============================================================================================
+    # Between this line ^^ and the one below this code is used to test...will delete
+    # this hack is here only to test the symmetrization
+    for prog in symm_programs:
+        prog.inst(II_definition)
+        prog.define_noisy_gate("II", [0, 1], append_kraus_to_gate(kraus_ops, II_mat))
+        prog.inst(II(0, 1))
+    # will delete code above here
+    # ============================================================================================
+
+    return symm_programs, flip_arrays
+
+
+def _consolidate_symmetrization_outputs(outputs: List[np.ndarray],
+                                        flip_arrays: List[Tuple[bool]]) -> np.ndarray:
+    """
+    Given bitarray results from a series of symmetrization programs, appropriately flip output
+    bits and consolidate results into new bitarrays.
+
+    :param outputs: a list of the raw bitarrays resulting from running a list of symmetrized
+        programs; for example, the results returned from _measure_bitstrings
+    :param flip_arrays: a list of boolean arrays in one-to-one correspondence with the list of
+        outputs indicating which qubits where flipped before each bitarray was measured.
+    :return: an np.ndarray consisting of the consolidated bitarray outputs which can be treated as
+        the symmetrized outputs of the original programs passed into a symmetrization method. See
+        estimate_observables for example usage.
+    """
+    assert len(outputs) == len(flip_arrays)
+
+    output = []
+    for bitarray, flip_array in zip(outputs, flip_arrays):
+        if len(flip_array) == 0:
+            output.append(bitarray)
+        else:
+            output.append(bitarray ^ flip_array)
+
+    return np.vstack(output)
+
+
+def _measure_bitstrings(qc, programs: List[Program], meas_qubits: List[int],
+                        num_shots: int = 600) -> List[np.ndarray]:
+    """
+    Wrapper for appending measure instructions onto each program, running the program,
+    and accumulating the resulting bitarrays.
+
+    :param qc: a quantum computer object on which to run each program
+    :param programs: a list of programs to run
+    :param meas_qubits: groups of qubits to measure for each program
+    :param num_shots: the number of shots to run for each program
+    :return: a len(programs) long list of num_shots by num_meas_qubits bit arrays of results for
+        each program.
+    """
+    results = []
+    for program in programs:
+        # copy the program so the original is not mutated
+        prog = program.copy()
+        ro = prog.declare('ro', 'BIT', len(meas_qubits))
+        for idx, q in enumerate(meas_qubits):
+            prog += MEASURE(q, ro[idx])
```

Comment:
so for `run_symmetrized_readout` users are expected to pass a MEASURE-less Program, but for `run` the MEASUREs need to be there?

Reply:
Yes. `run_symmetrized_readout` appends gates to the end of their program and then appends MEASUREs.  We should be comparing `run_and_measure` and `run_symmetrized_readout`.

If you would like, we could strip MEASUREs and then append the gates and re-append MEASUREs. But this behavior would be unlike `run` and unlike `run_and_measure`.